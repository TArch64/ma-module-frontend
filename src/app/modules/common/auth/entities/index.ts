export { User, IUserJSON } from './user';
export { InsecureUser, IInsecureUserJSON } from './insecure-user';
