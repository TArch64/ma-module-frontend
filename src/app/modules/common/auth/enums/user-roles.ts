export enum UserRoles {
    ADMIN = 'ADMIN',
    MENTOR = 'MENTOR',
    STUDENT = 'STUDENT'
}
