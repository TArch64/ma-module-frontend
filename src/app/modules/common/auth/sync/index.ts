export {CommonAuthSync, ICommonAuthSync, ISignInResponse} from './common-auth.sync';
