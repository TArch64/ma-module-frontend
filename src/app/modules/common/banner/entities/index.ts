export { IBannerOptions, BannerOptions } from './banner-options';
export { BannerRef } from './banner-ref';
export { BannerEventsProxy } from './banner-events-proxy';
