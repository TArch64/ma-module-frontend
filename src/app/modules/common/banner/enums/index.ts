export {BannerTypes} from './banner-types';
