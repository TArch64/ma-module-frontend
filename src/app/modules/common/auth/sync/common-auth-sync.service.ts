import {Injectable} from "@angular/core";
import {Observable} from "rxjs";
import {HttpClient} from "@angular/common/http";
import {IUserJSON} from "../entities";
import {ApiPathService} from "@common/core";

export interface ISignInResponse {
    token: string;
}

@Injectable({ providedIn: 'root' })
export class CommonAuthSyncService {
    constructor(
        private readonly apiPath: ApiPathService,
        private readonly httpClient: HttpClient
    ) {}

    public signIn(email: string, password: string): Observable<ISignInResponse> {
        const url = this.apiPath.build(['auth', 'sign-in']);
        return this.httpClient.post<ISignInResponse>(url, { email, password });
    }

    public loadCurrentUser(): Observable<IUserJSON> {
        const url = this.apiPath.build(['users', 'current']);
        return this.httpClient.get<IUserJSON>(url);
    }
}
