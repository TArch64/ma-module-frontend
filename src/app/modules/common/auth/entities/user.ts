import {UserRoles} from "../enums";

export interface IUserJSON {
    id: number;
    username: string;
    role: UserRoles;
}

export class User {
    public static fromJSON(json: IUserJSON): User {
        return new User(json.id, json.username, json.role);
    }

    private constructor(
        public readonly id: number,
        public readonly username: string,
        public readonly role: UserRoles
    ) {}
}
