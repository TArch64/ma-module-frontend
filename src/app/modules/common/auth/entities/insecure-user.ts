import { UserRoles } from '../enums';
import { IUserJSON, User } from './user';

export interface IInsecureUserJSON extends IUserJSON {
    email: string;
}

export class InsecureUser extends User {
    public static fromJSON(json: IInsecureUserJSON): InsecureUser {
        return new InsecureUser(
            json.id,
            json.username,
            json.role,
            InsecureUser.FAKE_AVATAR_URL,
            json.email
        );
    }

    private constructor(
        id: string,
        username: string,
        role: UserRoles,
        avatarUrl: string,
        public readonly email: string
    ) {
        super(id, username, role, avatarUrl);
    }

    public clone(overrides: Partial<IInsecureUserJSON> = {}): InsecureUser {
        return InsecureUser.fromJSON({ ...this, ...overrides });
    }
}
