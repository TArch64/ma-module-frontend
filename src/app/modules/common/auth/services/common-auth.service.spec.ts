import {Observable, of} from "rxjs";
import {IStorageService, IWindow} from "@common/core";
import {ICommonAuthSync, ISignInResponse} from '../sync';
import {IUserJSON} from "../entities";
import {UserRoles} from "../enums";
import {CommonAuthService} from './common-auth.service';

const TEST_TOKEN = 'test';

const createUserJson = (attrs: Partial<IUserJSON> = {}): IUserJSON => ({
    id: 'test',
    role: UserRoles.ADMIN,
    username: 'test',
    ...attrs
})

class MockCommonAuthSync implements ICommonAuthSync {
    public static create(): MockCommonAuthSync {
        return new MockCommonAuthSync();
    }

    mockLoadCurrentUser = jest.spyOn<ICommonAuthSync, 'loadCurrentUser'>(this, 'loadCurrentUser');

    signIn(email: string, password: string): Observable<ISignInResponse> {
        return of({token: TEST_TOKEN});
    }

    loadCurrentUser(): Observable<IUserJSON> {
        return of(createUserJson());
    }
}

class MockStorageService implements IStorageService {
    static create(): MockStorageService {
        return new MockStorageService();
    }

    readonly mockGetItem = jest.spyOn<IStorageService, 'getItem'>(this, 'getItem');
    readonly mockSetItem = jest.spyOn<IStorageService, 'setItem'>(this, 'setItem');
    readonly mockRemoveItem = jest.spyOn<IStorageService, 'removeItem'>(this, 'removeItem');

    getItem<T>(key: string): T | null {
        return null;
    }

    setItem<T>(key: string, value: T): void {}

    removeItem(key: string): void {}

    buildKey(key: string): string {
        return key;
    }
}

class MockWindow implements IWindow {
    static create(): MockWindow {
        return new MockWindow();
    }

    innerHeight = 1000;
    innerWidth = 1000;

    location = {
        reload: jest.fn()
    }

    addEventListener() {}
    removeEventListener() {}
}

function createService(sync: ICommonAuthSync, storageService: IStorageService, window: IWindow): CommonAuthService {
    return new CommonAuthService(storageService, sync, window);
}

describe('is signed in', () => {
    test('should return true if token in storage', () => {
        const storage = MockStorageService.create();
        storage.mockGetItem.mockReturnValue('test')

        const service = createService(MockCommonAuthSync.create(), storage, MockWindow.create());

        expect(service.isSignedIn).toBeTruthy();
    });

    test('should return false if no token in storage', () => {
        const storage = MockStorageService.create();
        storage.mockGetItem.mockReturnValue(null);

        const service = createService(MockCommonAuthSync.create(), storage, MockWindow.create());

        expect(service.isSignedIn).toBeFalsy();
    });
});

describe('sign in', () => {
    test('should save token', async () => {
        const storage = MockStorageService.create();
        const service = createService(MockCommonAuthSync.create(), storage, MockWindow.create());

        await service.signIn('test', 'test').toPromise();

        expect(storage.mockSetItem).toHaveBeenCalledWith(expect.any(String), TEST_TOKEN);
        expect(service.authToken).toEqual(TEST_TOKEN);
    });

    test('should update current user', async () => {
        const service = createService(MockCommonAuthSync.create(), MockStorageService.create(), MockWindow.create());

        await service.signIn('test', 'test').toPromise();

        expect(service.currentUserSnapshot).toMatchSnapshot();
    });
});

describe('sign out', () => {
    test('should remove token', () => {
        const storage = MockStorageService.create();
        const service = createService(MockCommonAuthSync.create(), storage, MockWindow.create());

        service.signOut();

        expect(storage.mockRemoveItem).toHaveBeenCalled();
    });

    test('should refresh window', () => {
        const window = MockWindow.create();
        const service = createService(MockCommonAuthSync.create(), MockStorageService.create(), window);

        service.signOut();

        expect(window.location.reload).toHaveBeenCalled();
    });
});

describe('fetch current user', () => {
    test('should load user if there no any', async () => {
        const sync = MockCommonAuthSync.create();
        const service = createService(sync, MockStorageService.create(), MockWindow.create());

        await service.fetchCurrentUser().toPromise();

        expect(sync.mockLoadCurrentUser).toHaveBeenCalled();
    });

    test('should get cached user', async () => {
        const sync = MockCommonAuthSync.create();
        const service = createService(sync, MockStorageService.create(), MockWindow.create());

        await service.signIn('test', 'test').toPromise();
        sync.mockLoadCurrentUser.mockClear();

        await service.fetchCurrentUser().toPromise();

        expect(sync.mockLoadCurrentUser).not.toHaveBeenCalled();
    });
});
