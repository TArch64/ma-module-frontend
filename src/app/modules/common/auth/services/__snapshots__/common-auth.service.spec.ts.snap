// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sign in should update current user 1`] = `
User {
  "avatarUrl": "https://avataaars.io/?avatarStyle=Circle&topType=WinterHat4&accessoriesType=Blank&hatColor=Red&facialHairType=BeardMajestic&facialHairColor=BlondeGolden&clotheType=BlazerShirt&eyeType=Default&eyebrowType=UnibrowNatural&mouthType=Default&skinColor=Brown",
  "id": "test",
  "role": "ADMIN",
  "username": "test",
}
`;
